### ISSUE TITLE
On 06-5e-03 20200609 was good, but 20200616 removes SRBDS mitigation.

### ISSUE THREAD

**Issue by zarniwhoop73** [intel/Intel-Linux-Processor-Microcode-Data-Files#36]
I have an i3  which was working ok with the early-loaded  20200616 microcode (0xdc), so I assumed that the issues with 06-4e-03 for that version were specific to -4e-. With linux-5.7.2 I have Microcode listed as the mitigation for srbds.

But the 20200616 release downgrades _both_ skylake variants to the microcode from last year and therefore loses the mitigation.

**Comment by esyr-rh** [comment]
On Tue, Jun 16, 2020 at 06:20:35PM -0700, Ken Moffat wrote:
> I have an i3  which was working ok with the early-loaded  20200616 microcode (0xdc), so I assumed that the issues with 06-4e-03 for that version were specific to -4e-. With linux-5.7.2 I have Microcode listed as the mitigation for srbds.
> 
> But the 20200616 release downgrades _both_ skylake variants to the microcode from last year and therefore loses the mitigation.

See [1].

[1] https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/31#issuecomment-644885826

**Comment by hmh** [comment]
Yes, whatever issue can happen with 0x506e3 does not happen to everyone.  We also had at least one good report (as in: it works) in Debian for 0x506e3.

**Comment by eworm-de** [comment]
I had a report for Arch Linux as well....

**Comment by hmh** [comment]
@eworm-de: was the report good (20200609 works on 0x506e3) or bad (20200609 causes issues on 0x506e3) ?

**Comment by hmh** [comment]
Anyway, it clearly can cause trouble on some systems: someone just posted a confirmation of a hang on their 0x506e3 (Xeon E3) system to the issue dealing with 0x406e3[1]

[1] https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/31#issuecomment-645181792

**Comment by eworm-de** [comment]
20200609 caused issues:
https://bugs.archlinux.org/task/67011

**Comment by jelearn** [comment]
With regards to the [Special Register Buffer Data Sampling Advisory](https://www.intel.com/content/www/us/en/security-center/advisory/intel-sa-00320.html) and the [Intel Microcode released on GitHub](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files) there appears to be a regression introduced that affects the following Intel processor:

skylake
Version: Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
Signature: Type 0, Family 6, Model 78, Stepping 3
Microcode patch: [06-4e-03](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/blob/main/intel-ucode/06-4e-03)

From what I've read ([from Ubuntu](https://launchpad.net/ubuntu/+source/intel-microcode/3.20200609.0ubuntu0.20.04.2/) and this similar issues here in GitHub) that their fix prevents some systems from booting and so was reverted.

In my case late loading of the microcode did not work either, however early loading did (as discussed below).

The following outlines what I did to address the issue to the best of my ability.  It assumes a local clone of the `Intel-Linux-Processor-Microcode-Data-Files` repo.

The "microcode-20200616" release of that processor's microcode patch (which has made it into Linux distributions such as [Fedora](https://koji.fedoraproject.org/koji/buildinfo?buildID=1631760) and [Ubuntu 20.04](https://launchpad.net/ubuntu/+source/intel-microcode/3.20200609.0ubuntu0.20.04.2/)) results in the SRBDS mitigation to be disabled:

```
$ cat /sys/devices/system/cpu/vulnerabilities/srbds && dmesg | grep microcode
Vulnerable: No microcode
[    0.000000] microcode: microcode updated early to revision 0xd6, date = 2019-10-03
[    0.119201] SRBDS: Vulnerable: No microcode
[    0.678330] microcode: sig=0x406e3, pf=0x80, revision=0xd6
[    0.678355] microcode: Microcode Update Driver: v2.2.
```

Specifically, by default the following microcode is chosen in that release, which has already been identified as being vulnerable:

```
$ /usr/sbin/iucode_tool -tb -lS ~/git/Intel-Linux-Processor-Microcode-Data-Files/intel-ucode/* | tail -n -1
/usr/sbin/iucode_tool: system has processor(s) with signature 0x000406e3
  066/001: sig 0x000406e3, pf_mask 0xc0, 2019-10-03, rev 0x00d6, size 101376
```

To compare with the previous "iucode_tool" output, you can see that the "microcode-20200609" release (`git checkout microcode-20200609`) results in the proper microcode being chosen to address the SRBDS vulnerability (according to the original release notes).

```
$ /usr/sbin/iucode_tool -tb -lS ~/git/Intel-Linux-Processor-Microcode-Data-Files/intel-ucode/* | tail -n -1
/usr/sbin/iucode_tool: system has processor(s) with signature 0x000406e3
  066/001: sig 0x000406e3, pf_mask 0xc0, 2020-04-27, rev 0x00dc, size 104448
```

Reverting to the previous microcode in the "microcode-20200609" release that was published to address the vulnerability results in the microcode being loaded properly in my case.

The following steps (after checking out the "microcode-20200609" release) will most likely correct the issue (on Ubuntu 20.04).
Before doing this you need to feel comfortable with being able to revert the change in case on your platform it does not work.

**_These instructions come with no warranty, use at your own risk_**.

It will only work with the processor above (i.e. if the `iucode_tool` steps above produce the same output).

Line overview:

1. Clone the Intel Linux Microcode Git Hub repo
2. Checkout the previous release that has the SRBDS vulnerability fix.
3. Backup the current microcode (in case you need to revert).
4. Copy the microcode to the system directory
5. Rename it to have `.initramfs` to be loaded early (without this, it will not work)
6. Check that you have indeed updated the microcode as expected.

Commands:

```
mkdir ~/git && cd ~/git && git clone https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files.git && \
cd ~/git/Intel-Linux-Processor-Microcode-Data-Files  && git checkout microcode-20200609 && \
sudo /lib/firmware/intel-ucode/06-4e-03 /root/. && \
sudo cp ~/git/Intel-Linux-Processor-Microcode-Data-Files/intel-ucode/06-4e-03 /lib/firmware/intel-ucode/06-4e-03 && \
sudo mv /lib/firmware/intel-ucode/06-4e-03 /lib/firmware/intel-ucode/06-4e-03.initramfs && \
/usr/sbin/iucode_tool -tb -lS /lib/firmware/intel-ucode/*
```

The output should be like:

```
/usr/sbin/iucode_tool: system has processor(s) with signature 0x000406e3
  066/001: sig 0x000406e3, pf_mask 0xc0, 2020-04-27, rev 0x00dc, size 104448
```

If you are satisfied, update your `initramfs` image to be loaded at next boot as shown below:

```
sudo update-initramfs -u && echo SUCCESS || echo ERROR
```

If successful, reboot.

Result (after reboot):

```
$ cat /sys/devices/system/cpu/vulnerabilities/srbds && \
> dmesg | grep microcode
Mitigation: Microcode
[    0.000000] microcode: microcode updated early to revision 0xdc, date = 2020-04-27
[    0.684607] microcode: sig=0x406e3, pf=0x80, revision=0xdc
[    0.684635] microcode: Microcode Update Driver: v2.2.
```

If this did not work for you, reboot (holding F1 as it starts) and choose the "recovery" boot option from the GRUB menu.  Choose the option to start a root shell.  Remove the file that was added, copy the backup in to the same directory, and re-run the `update-initramfs -u` command.  Once complete, reboot and hope a fix is released some time in the future.

No other CPUs have been tested to determine if the microcode has been reverted and thus re-enabling the vulnerability.

I think that if one were to follow the same diagnostic steps to identify which microcode is related to your CPU may lead to a similar outcome if that CPU has the vulnerability.  The steps may need to be adjusted per Linux Distribution.

This seems to be related to #38 but there is no news from Intel.

**Comment by esyr-rh** [comment]
06-5e-03 microcode is updated to revision 0xe2 (that contains SRBDS mitigations) with microcode-20201110 release.

**Comment by hmh** [comment]
Since 0x406e3 boothang is reportedly fixed, the one with 0x506e3 is likely fixed as well and it contains the SRBDS mitigations.

It would be nice to get direct confirmation that the new microcode works on 0x506e3, though.

**Comment by zarniwhoop73** [comment]
0xe2 works on my 06-5e-03 (i3 skylake), but then the old pre-revert version did too.

**Comment by hmh** [comment]
Yeah, we'd need testing from someone who had issues with the pre-revert microcode update :-(  And now some complains of failure-to-reboot for 0x406e3 with the newest update surfaced, so it might not have been fixed (or some bad cases still remain)...

**Comment by hmh** [comment]
This is related to #31 

Systems with 06-4e-03 and 06-5e-03 processors where the firmware has a microcode revision older than 0x80 *cannot* be safely updated to revisions 0xd8 and above, and may hang.

As usual "update your firmware" if such an update is available.  If the microcode revision in the firmware is newer than 0x80, it should be possible to use the microcode updates in the latest Intel release (20201118 as I write this).

**Comment by esyr-rh** [comment]
New revision 0xea of 06-5e-03 microcode file has been published as part of [microcode-20210608](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/releases/tag/microcode-20210608) release, it may be worth to try it out.
